processor/queuedprocessor/queued_processor.go

```diff
@@ -39,53 +40,156 @@ type queuedSpanProcessor struct {
 	name                     string
 	queue                    *queue.BoundedQueue
 	logger                   *zap.Logger
-	sender                   consumer.TraceConsumer
+	traceNext                consumer.TraceConsumer
+	metricNext               consumer.MetricsConsumer
 	numWorkers               int
 	retryOnProcessingFailure bool
 	backoffDelay             time.Duration
-	statsTags                []tag.Mutator
 	stopCh                   chan struct{}
 	stopOnce                 sync.Once
 }
 
 var _ consumer.TraceConsumer = (*queuedSpanProcessor)(nil)
+var errorRefused = errors.New("failed to add to the queue")
+
+type queueItem interface {
+	context() context.Context
+	queuedTime() time.Time
+	export(sp *queuedSpanProcessor) error
+	onAccepted()
+	onRefused(logger *zap.Logger, err error)
+	onDropped(logger *zap.Logger, err error)
+}
+
+type baseQueueItem struct {
+	ctx context.Context
+	qt  time.Time
+}
 
-type queueItem struct {
-	queuedTime     time.Time
+func (item *baseQueueItem) context() context.Context {
+	return item.ctx
+}
+
+func (item *baseQueueItem) queuedTime() time.Time {
+	return item.qt
+}
+
+type traceQueueItem struct {
+	baseQueueItem
 	td             pdata.Traces
 	spanCountStats *processor.SpanCountStats
-	ctx            context.Context
+}
+
+func newTraceQueueItem(ctx context.Context, td pdata.Traces, instanceName string) queueItem {
+	// Ensure always have the TagProcessorNameKey set in the context.
+	// TODO: Remove this after #1312 is merged and verify.
+	ctx, _ = tag.New(ctx, tag.Upsert(processor.TagProcessorNameKey, instanceName))
+	return &traceQueueItem{
+		baseQueueItem:  baseQueueItem{ctx: ctx, qt: time.Now()},
+		td:             td,
+		spanCountStats: processor.NewSpanCountStats(td),
+	}
+}
+
+func (item *traceQueueItem) onAccepted() {
+	processor.RecordsSpanCountMetrics(item.ctx, item.spanCountStats, processor.StatReceivedSpanCount)
+	obsreport.ProcessorTraceDataAccepted(item.ctx, item.spanCountStats.GetAllSpansCount())
+}
+
+func (item *traceQueueItem) onRefused(logger *zap.Logger, err error) {
+	// Count the StatReceivedSpanCount even if items were refused.
+	processor.RecordsSpanCountMetrics(item.ctx, item.spanCountStats, processor.StatReceivedSpanCount)
+	// TODO: in principle this may not end in data loss because this can be
+	// in the same call stack as the receiver, ie.: the call from the receiver
+	// to here is synchronous. This means that actually it could be proper to
+	// record this as "refused" instead of "dropped".
+	//
+	// Right now keep the previous behavior to not change too many things in this PR.
```

Comment:
Can we just change this to record refused and remove this comment?

Reply:
Done.